### PR TITLE
Load Speed Increase

### PR DESCRIPTION
**Pull request by timsnow**
The speed of loading all datasets has been increased, in the case of large datasets dramatically so.

In addition, if an Idata or Tdata node contains comma delimited data, as opposed to element delimited, this will handled nicely and loaded as well.

### CODE REVIEW

**Comment by timsnow** [comment]
From Travis:

Result for utest_nxcansas_writer (/home/travis/build/SasView/sasview/test/fileconverter/test): FAILED
Running tests...
----------------------------------------------------------------------
E.
======================================================================
ERROR [0.210s]: test_write_1d (tests.nxcansas_writer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "utest_nxcansas_writer.py", line 39, in test_write_1d
    self._check_metadata(data, self.data_1d)
  File "utest_nxcansas_writer.py", line 54, in _check_metadata
    self.assertAlmostEqual(written.sample.thickness, correct.sample.thickness)
  File "/home/travis/miniconda/lib/python2.7/unittest/case.py", line 554, in assertAlmostEqual
    if round(abs(second-first), places) == 0:
TypeError: ufunc 'subtract' did not contain a loop with signature matching types dtype('S32') dtype('S32') dtype('S32')
----------------------------------------------------------------------
Ran 2 tests in 0.390s
FAILED (errors=1)
Generating XML reports...
=== testing: /home/travis/build/SasView/sasview/test/fileconverter/test/utest_nxcansas_writer.py
Result for utest_corfunc: SUCCESS
Result for utest_manipulations: SUCCESS
Result for utest_abs_reader: SUCCESS
Result for utest_cansas (/home/travis/build/SasView/sasview/test/sasdataloader/test): FAILED
Running tests...
----------------------------------------------------------------------
..FF.F16:55:38 - INFO : sas.sascalc.dataloader.readers.xml_reader:  75: Opening and ending tag mismatch: c line 1 and b, line 1, column 11
.....F.
======================================================================
FAIL [0.020s]: test_cansas_xml (tests.cansas_reader_xml)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "utest_cansas.py", line 126, in test_cansas_xml
    self._check_data(dataloader[i])
  File "utest_cansas.py", line 96, in _check_data
    self.assertTrue(data.x.size == 138)
AssertionError: False is not true
======================================================================
FAIL [0.014s]: test_double_trans_spectra (tests.cansas_reader_xml)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "utest_cansas.py", line 145, in test_double_trans_spectra
    self._check_data(item)
  File "utest_cansas.py", line 96, in _check_data
    self.assertTrue(data.x.size == 138)
AssertionError: False is not true
======================================================================
FAIL [0.006s]: test_invalid_cansas (tests.cansas_reader_xml)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "utest_cansas.py", line 179, in test_invalid_cansas
    self.assertTrue(data.x.size == 2)
AssertionError: False is not true
======================================================================
FAIL [0.014s]: test_save_cansas_v1_0 (tests.cansas_reader_xml)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "utest_cansas.py", line 210, in test_save_cansas_v1_0
    self._check_data(dataloader[i])
  File "utest_cansas.py", line 96, in _check_data
    self.assertTrue(data.x.size == 138)
AssertionError: False is not true
----------------------------------------------------------------------
Ran 13 tests in 0.146s
FAILED (failures=4)
Generating XML reports...
=== testing: /home/travis/build/SasView/sasview/test/sasdataloader/test/utest_cansas.py
Result for utest_red2d_reader: SUCCESS
Result for utest_averaging: SUCCESS
Result for utest_ascii: SUCCESS
Result for utest_use_cases (/home/travis/build/SasView/sasview/test/sasinvariant/test): FAILED
Running tests...
----------------------------------------------------------------------
..F............
======================================================================
FAIL [0.043s]: test_use_case_3 (tests.TestInvPinholeSmear)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "utest_use_cases.py", line 324, in test_use_case_3
    self.assertAlmostEquals(v, 0.117226896,2)
AssertionError: 0.11066542072334296 != 0.117226896 within 2 places
----------------------------------------------------------------------
Ran 15 tests in 0.292s
FAILED (failures=1)
Generating XML reports...
=== testing: /home/travis/build/SasView/sasview/test/sasinvariant/test/utest_use_cases.py
Result for utest_data_handling (/home/travis/build/SasView/sasview/test/sasinvariant/test): FAILED
Running tests...
----------------------------------------------------------------------
.............F........
======================================================================
FAIL [0.023s]: test_qstar_low_q_power_law (tests.TestInvariantCalculator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "utest_data_handling.py", line 201, in test_qstar_low_q_power_law
    self.assertAlmostEqual(inv._low_extrapolation_function.power, 0.0655, 3)
AssertionError: 0.050190957216523308 != 0.0655 within 3 places
----------------------------------------------------------------------
Ran 22 tests in 0.372s
FAILED (failures=1)
Generating XML reports...
=== testing: /home/travis/build/SasView/sasview/test/sasinvariant/test/utest_data_handling.py
----------------------------------------------
Results by test modules:
    PASSED: 11
    FAILED: 4    (27%)
Results by tests:
    Tests run:    147
    Tests failed: 6
    Test errors:  1
----------------------------------------------

**Comment by timsnow** [comment]
Error report from Travis attached (in case it is deleted):

In essence there appears to be 3 main errors:

1) test.nxcansas_writer:

  File "/home/travis/miniconda/lib/python2.7/unittest/case.py", line 554, in assertAlmostEqual
    if round(abs(second-first), places) == 0:
TypeError: ufunc 'subtract' did not contain a loop with signature matching types dtype('S32') dtype('S32') dtype('S32')


2) tests.cansas_reader_xml (multiple occurrences):

  File "utest_cansas.py", line 96, in _check_data
    self.assertTrue(data.x.size == 138)
AssertionError: False is not true


3) tests.TestInvariantCalculator / tests.TestInvPinholeSmear:

  File "utest_use_cases.py", line 324, in test_use_case_3
    self.assertAlmostEquals(v, 0.117226896,2)
AssertionError: 0.11066542072334296 != 0.117226896 within 2 places

  File "utest_data_handling.py", line 201, in test_qstar_low_q_power_law
    self.assertAlmostEqual(inv._low_extrapolation_function.power, 0.0655, 3)
AssertionError: 0.050190957216523308 != 0.0655 within 3 places


I'm going to investigate 1 & 2, however, I'm not totally sure about 3. Perhaps the data isn't being loaded in suitably for the invariant calculation and this is causing a problem, if not perhaps the big merge from master has introduced something new...



[2017-04-11 - Travis Output.txt](https://github.com/SasView/sasview/files/912523/2017-04-11.-.Travis.Output.txt)

**Comment by timsnow** [comment]
I now understand all three errors and will work on this when I get back as there's some extra magic going on...

File: src/sas/sascalc/dataloader/readers/cansas_reader.py [hunk 1]
```diff
@@ -196,303 +196,477 @@ def _parse_entry(self, dom, recurse=False):
...
+                for setupNode in sasNode[0]:
+                    if setupNode.text is not None:
+                        # Iterating through the tags in the unit node, getting their tag name and respective unit
+                        setupTagName = setupNode.tag.replace(self.base_ns, "")
+                        units = setupNode.attrib.get("unit", "")
+
+                        # Creating our data array first, if there's only one dataNode we will handle this...
+                        startArray = np.fromstring(setupNode.text, dtype=float, sep=",")
+                        
+                        if multipleEntries == True:
```
Comment (andyfaff) [inline src/sas/sascalc/dataloader/readers/cansas_reader.py]: `if multiple_entries:`

File: src/sas/sascalc/dataloader/readers/cansas_reader.py [hunk 2]
```diff
@@ -196,303 +196,477 @@ def _parse_entry(self, dom, recurse=False):
...
+                            elif dataTagName == "dQw":
+                                self.current_dataset.dxw[loopIter] = dataArray
+                            elif dataTagName == "dQl":
+                                self.current_dataset.dxl[loopIter] = dataArray
+
+                self._check_for_empty_resolution()
+                self.data.append(self.current_dataset)
+
+            # If it's not data, let's check for other tags starting with skippable ones...
+            elif currentTagName == "fitting_plug_in" or currentTagName == "pr_inversion" or currentTagName == "invariant":
```
Comment (andyfaff) [inline src/sas/sascalc/dataloader/readers/cansas_reader.py]: `elif currentTagName in ['fitting_plug_in', 'pr_inversion', 'invariant']:`

**Comment by andyfaff** [comment]
This file loader is pretty big. Are there improvements possible from use of [XPath](https://en.wikipedia.org/wiki/XPath)? For example, loops can be removed by creating an Xpath query to get all the Idata nodes for the first dataset, etc.

File: src/sas/sascalc/dataloader/readers/cansas_reader.py [hunk 3]
```diff
@@ -196,303 +196,477 @@ def _parse_entry(self, dom, recurse=False):
...
+            # Get the element name and set the current name's level
+            currentTagName = sasNode.tag.replace(self.base_ns, "")
+            # As this is the most likely tag to examine, lets put it first!
+            if currentTagName == "SASdata":
+                # Are there multiple entries here?
+                if len(sasNode) <= 1:
+                    multipleEntries = False
+                else:
+                    multipleEntries = True
+
```
Comment (pkienzle) [inline src/sas/sascalc/dataloader/readers/cansas_reader.py]: multipleEntries = (len(sasNode) <= 1)

**Comment by pkienzle** [comment]
Closing pull request for now.  Reopen it when you are ready for others to look at it.

**Pull request closed without merging**